### PR TITLE
Improve vmdk_ops_test 

### PR DESCRIPTION
**Pull request by ashahi1**
Fix for issue #832 

Tested the CI tests with one esx and two photon os VMs.

### CODE REVIEW

**Comment by shuklanirdesh82** [comment]
@ashahi1  can you please run vmdk_ops_test when there is only vsan datastore present in the inventory?

File: esx_service/vmdk_ops_test.py [hunk 1]
```diff
@@ -39,6 +39,7 @@
 import error_code
 import vmdk_utils
 import random
+import time
```
Comment (shuklanirdesh82) [inline esx_service/vmdk_ops_test.py]: I guess we don't need this. Please remove.

File: esx_service/vmdk_ops_test.py [hunk 2]
```diff
@@ -731,7 +730,7 @@ def setUp(self):
                 self.datastore_path = datastore[2]
                 logging.debug("datastore_name=%s datastore_path=%s", self.datastore_name,
                                                                      self.datastore_path)
-                if len(datastores) >= 1:
+                if len(datastores) >1:
```
Comment (kerneltime) [inline esx_service/vmdk_ops_test.py]: Why is this change needed?

File: esx_service/vmdk_ops_test.py (hunk 1 above)
Comment (ashahi1) [inline esx_service/vmdk_ops_test.py]: I missed it. Will drop it. Thanks.

File: esx_service/vmdk_ops_test.py (hunk 2 above)
Comment (ashahi1) [inline esx_service/vmdk_ops_test.py]: old code: 
if len(datastores) >=1:
                    datastore1 = datastores[1]    >>>>>> this will cause failure if we have only one datastore in testbed
                    self.datastore1_name = datastore1[0]
                    self.datastoer1_path = datastore[2]
                    logging.debug("Found second datastore: datastore_name=%s datastore_path=%s",
                                  self.datastore1_name, self.datastore1_path)


This new condition enables that code within if block will only be executed if we have more than one datastore.

File: esx_service/vmdk_ops_test.py [hunk 3]
```diff
@@ -1086,23 +1096,32 @@ def test_vmdkops_on_different_tenants(self):
...
+        result = vmdk_ops.executeRequest(vm2_uuid, self.vm2_name, self.vm2_config_path, 'list', None, opts)        
         # listVMDK return vol name with the format like vol@datastore
         self.assertEqual(3, len(result))
-        self.assertEqual("tenant2_vol1@"+self.datastore_name, result[0]['Name'])
-        self.assertEqual("tenant2_vol2@"+self.datastore_name, result[1]['Name'])
-        self.assertEqual("tenant2_vol3@"+self.datastore_name, result[2]['Name'])
-    
+         
+        volume_names = ["tenant2_vol1@"+self.datastore_name,"tenant2_vol2@"+self.datastore_name,"tenant2_vol3@"+self.datastore_name]
+        for name in volume_names:
```
Comment (shuklanirdesh82) [inline esx_service/vmdk_ops_test.py]: It would be better to move this piece of code into verification util instead of copying multiple places. 

Please address the comment into your next diff.

File: esx_service/vmdk_ops_test.py [hunk 4]
```diff
@@ -100,7 +100,8 @@ class VmdkCreateRemoveTestCase(unittest.TestCase):
     valid_attach_as_opt_1 = {volume_kv.ATTACH_AS: u'persistent'}
     valid_attach_as_opt_2 = {volume_kv.ATTACH_AS: u'independent_persistent'}
     name = ""
-    vm_name = 'test-vm'
+    random_id = random.randint(0, 65536)
```
Comment (msterin) [inline esx_service/vmdk_ops_test.py]: please separate this into a `generate_test_vm_name` function instead of cut-n-paste throughout the code. 
Also, replace these 2 lines in all places below the function call please

File: esx_service/vmdk_ops_test.py [hunk 5]
```diff
@@ -445,6 +447,18 @@ def remove_vm(si, vm):
         task = vm.Destroy_Task()
         vmdk_ops.wait_for_tasks(si, [task])
 
+def checkIfVolumeExist(volume_names,result):
+    #checks if volume exists in the list of vmdks present in result
+    if result:
```
Comment (msterin) [inline esx_service/vmdk_ops_test.py]: this would be much better (and less nesting) if it said:
```python
if not result:
   return True

for name in volume_names:
    ...   
```
also I do not understnad the login... `if not result` returns true in the code. Why ?

File: esx_service/vmdk_ops_test.py [hunk 6]
```diff
@@ -731,7 +740,7 @@ def setUp(self):
                 self.datastore_path = datastore[2]
                 logging.debug("datastore_name=%s datastore_path=%s", self.datastore_name,
                                                                      self.datastore_path)
-                if len(datastores) >= 1:
+                if len(datastores) >1:
```
Comment (msterin) [inline esx_service/vmdk_ops_test.py]: So if we have only one datastore, `datastore1`, `self.datastore1_name ` and  `self.datastoer1_path ` will be not set. Is it OK ? '
'also - nit: space after `>` is missing

File: esx_service/vmdk_ops_test.py [hunk 7]
```diff
@@ -972,9 +981,8 @@ def test_vmdkops_on_tenant_vm(self):
         # "result"" should have two volumes :tenant1_vol2@default_datastore, and  tenant1_vol3@datastore
         # these volumes are created at default_datastore, which is self.datastore_name
         self.assertEqual(2, len(result))
-        self.assertEqual("tenant1_vol2@"+self.datastore_name, result[0]['Name'])
-        self.assertEqual("tenant1_vol3@"+self.datastore_name, result[1]['Name'])
-
+        volume_names = ["tenant1_vol2@"+self.datastore_name,"tenant1_vol3@"+self.datastore_name]
```
Comment (msterin) [inline esx_service/vmdk_ops_test.py]: spaces are missing around `+` and after `,`. Please stay with with the code style: https://www.python.org/dev/peps/pep-0008/

File: esx_service/vmdk_ops_test.py [hunk 8]
```diff
@@ -996,9 +1004,8 @@ def test_vmdkops_on_tenant_vm(self):
         # listVMDK return vol name with the format like vol@datastore
         # result should have two volumes :tenant1_vol2, and  tenant1_vol3
         self.assertEqual(2, len(result))
-        self.assertEqual("tenant1_vol2@"+self.datastore_name, result[0]['Name'])
-        self.assertEqual("tenant1_vol3@"+self.datastore_name, result[1]['Name'])
-
+        volume_names = ["tenant1_vol2@"+self.datastore_name,"tenant1_vol3@"+self.datastore_name]
```
Comment (msterin) [inline esx_service/vmdk_ops_test.py]: please stay with the code style. 

Also, I suggest a function for constructing volume_names - instead of the copy-pasted code used now

File: esx_service/vmdk_ops_test.py [hunk 9]
```diff
@@ -1086,23 +1093,19 @@ def test_vmdkops_on_different_tenants(self):
...
         opts = {}
         result = vmdk_ops.executeRequest(vm1_uuid, self.vm1_name, self.vm1_config_path, 'list', None, opts)
-       
         # listVMDK return vol name with the format like vol@datastore
         self.assertEqual(3, len(result))
-        self.assertEqual("tenant1_vol1@"+self.datastore_name, result[0]['Name'])
-        self.assertEqual("tenant1_vol2@"+self.datastore_name, result[1]['Name'])
-        self.assertEqual("tenant1_vol3@"+self.datastore_name, result[2]['Name'])
 
+        volume_names = ["tenant1_vol1@"+self.datastore_name,"tenant1_vol2@"+self.datastore_name,"tenant1_vol3@"+self.datastore_name]
```
Comment (msterin) [inline esx_service/vmdk_ops_test.py]: ditto

File: esx_service/vmdk_ops_test.py [hunk 10]
```diff
@@ -1086,23 +1093,19 @@ def test_vmdkops_on_different_tenants(self):
...
-       
+        result = vmdk_ops.executeRequest(vm2_uuid, self.vm2_name, self.vm2_config_path, 'list', None, opts)        
         # listVMDK return vol name with the format like vol@datastore
-        self.assertEqual(3, len(result))
-        self.assertEqual("tenant2_vol1@"+self.datastore_name, result[0]['Name'])
-        self.assertEqual("tenant2_vol2@"+self.datastore_name, result[1]['Name'])
-        self.assertEqual("tenant2_vol3@"+self.datastore_name, result[2]['Name'])
-    
+        self.assertEqual(3, len(result)) 
+        volume_names = ["tenant2_vol1@"+self.datastore_name,"tenant2_vol2@"+self.datastore_name,"tenant2_vol3@"+self.datastore_name]
```
Comment (msterin) [inline esx_service/vmdk_ops_test.py]: ditto - code style and less copy-n-paste if possible

File: esx_service/vmdk_ops_test.py (hunk 4 above)
Comment (ashahi1) [inline esx_service/vmdk_ops_test.py]: Created a separate function - generate_test_vm_name().

File: esx_service/vmdk_ops_test.py (hunk 5 above)
Comment (ashahi1) [inline esx_service/vmdk_ops_test.py]: Fixed it.

File: esx_service/vmdk_ops_test.py (hunk 6 above)
Comment (ashahi1) [inline esx_service/vmdk_ops_test.py]: Ran the vmdk_ops_test.py test only with one datastore - vmfs. Results are as follows:
================= 

Running unit tests in /tmp/vmdk_ops_unittest3474/vmdk_ops_test.py...
WARNING:root:dockvols creation failed - Operation not permitted on local.3-0
WARNING:root:dockvols creation failed - Operation not permitted on local.3-1
ss........sss....
----------------------------------------------------------------------
Ran 17 tests in 204.492s

OK (skipped=5)

File: esx_service/vmdk_ops_test.py (hunk 10 above)
Comment (ashahi1) [inline esx_service/vmdk_ops_test.py]: Created a separate function for volume_names - generate_volume_names
Fixed the code style too

File: esx_service/vmdk_ops_test.py [hunk 11]
```diff
@@ -972,9 +992,8 @@ def test_vmdkops_on_tenant_vm(self):
         # "result"" should have two volumes :tenant1_vol2@default_datastore, and  tenant1_vol3@datastore
         # these volumes are created at default_datastore, which is self.datastore_name
         self.assertEqual(2, len(result))
-        self.assertEqual("tenant1_vol2@"+self.datastore_name, result[0]['Name'])
-        self.assertEqual("tenant1_vol3@"+self.datastore_name, result[1]['Name'])
-
+        volume_names = ["tenant1_vol2@" + self.datastore_name, "tenant1_vol3@" + self.datastore_name]
```
Comment (shuklanirdesh82) [inline esx_service/vmdk_ops_test.py]: don't we need to use ```generate_volume_names``` here?

File: esx_service/vmdk_ops_test.py [hunk 12]
```diff
@@ -996,9 +1015,8 @@ def test_vmdkops_on_tenant_vm(self):
         # listVMDK return vol name with the format like vol@datastore
         # result should have two volumes :tenant1_vol2, and  tenant1_vol3
         self.assertEqual(2, len(result))
-        self.assertEqual("tenant1_vol2@"+self.datastore_name, result[0]['Name'])
-        self.assertEqual("tenant1_vol3@"+self.datastore_name, result[1]['Name'])
-
+        volume_names = ["tenant1_vol2@" + self.datastore_name, "tenant1_vol3@" + self.datastore_name]
```
Comment (shuklanirdesh82) [inline esx_service/vmdk_ops_test.py]: same as above

File: esx_service/vmdk_ops_test.py [hunk 13]
```diff
@@ -1086,23 +1104,21 @@ def test_vmdkops_on_different_tenants(self):
...
         opts = {}
         result = vmdk_ops.executeRequest(vm1_uuid, self.vm1_name, self.vm1_config_path, 'list', None, opts)
-       
         # listVMDK return vol name with the format like vol@datastore
         self.assertEqual(3, len(result))
-        self.assertEqual("tenant1_vol1@"+self.datastore_name, result[0]['Name'])
-        self.assertEqual("tenant1_vol2@"+self.datastore_name, result[1]['Name'])
-        self.assertEqual("tenant1_vol3@"+self.datastore_name, result[2]['Name'])
 
+        # volume_names = ["tenant1_vol1@"+self.datastore_name,"tenant1_vol2@"+self.datastore_name,"tenant1_vol3@"+self.datastore_name]
```
Comment (shuklanirdesh82) [inline esx_service/vmdk_ops_test.py]: need to remove dead code here

File: esx_service/vmdk_ops_test.py [hunk 14]
```diff
@@ -1086,23 +1104,21 @@ def test_vmdkops_on_different_tenants(self):
...
-       
+        result = vmdk_ops.executeRequest(vm2_uuid, self.vm2_name, self.vm2_config_path, 'list', None, opts)        
         # listVMDK return vol name with the format like vol@datastore
-        self.assertEqual(3, len(result))
-        self.assertEqual("tenant2_vol1@"+self.datastore_name, result[0]['Name'])
-        self.assertEqual("tenant2_vol2@"+self.datastore_name, result[1]['Name'])
-        self.assertEqual("tenant2_vol3@"+self.datastore_name, result[2]['Name'])
-    
+        self.assertEqual(3, len(result)) 
+        # volume_names = ["tenant2_vol1@"+self.datastore_name,"tenant2_vol2@"+self.datastore_name,"tenant2_vol3@"+self.datastore_name]
```
Comment (shuklanirdesh82) [inline esx_service/vmdk_ops_test.py]: same as above

File: esx_service/vmdk_ops_test.py [hunk 15]
```diff
@@ -86,6 +86,12 @@ def test_name_parse(self):
             except vmdk_ops.ValidationError as ex:
                 self.assertFalse(expected_result, "Expected vol name parsing to succeed for '{0}'"
                                  .format(full_name))
+    
+def generate_test_vm_name():
```
Comment (shuklanirdesh82) [inline esx_service/vmdk_ops_test.py]: nit: better to have method level doc/description that talks about the purpose of this helper

File: esx_service/vmdk_ops_test.py [hunk 16]
```diff
@@ -445,12 +451,34 @@ def remove_vm(si, vm):
         task = vm.Destroy_Task()
         vmdk_ops.wait_for_tasks(si, [task])
 
+def generate_volume_names(tenant, datastore_name):
```
Comment (shuklanirdesh82) [inline esx_service/vmdk_ops_test.py]: nit: need method's doc/description

File: esx_service/vmdk_ops_test.py [hunk 17]
```diff
@@ -445,12 +451,34 @@ def remove_vm(si, vm):
...
+    """ Returns a list of volume names """
+    volNames = []
+    for x in range(3):
+        volNames.append(tenant + "_vol" + str(x + 1) + "@" + datastore_name)
+    
+    return volNames
+    
+def checkIfVolumeExist(volume_names,result):
```
Comment (shuklanirdesh82) [inline esx_service/vmdk_ops_test.py]: same as above

File: esx_service/vmdk_ops_test.py [hunk 18]
```diff
@@ -445,12 +451,34 @@ def remove_vm(si, vm):
...
+    #checks if volume exists in the list of vmdks present in result
+    if not result:
```
Comment (msterin) [inline esx_service/vmdk_ops_test.py]: nit: the code is OK, but IMO it would be nicer to say 
```python
  if not result or not volume_names:
   return False
```

File: esx_service/vmdk_ops_test.py (hunk 11 above)
Comment (ashahi1) [inline esx_service/vmdk_ops_test.py]: @shuklanirdesh82  Test 'test_vmdkops_on_tenant_vm' creates a tenant and then attaches volume tenant1_vol1_name and then detaches the volume. And then finally checks that 'result' should have two volumes :tenant1_vol2@default_datastore, and tenant1_vol3@datastore. That's why I didn't use this method here. I have used this method at instances where code checks for all the three volumes. Suggestions?

Comment (shuklanirdesh82) [inline esx_service/vmdk_ops_test.py]: This sounds confusing. By looking at the tests I can confirm that test is deleting the first created volume ```tenant1_vol1@``` hence the verification only for the rest two volumes. 

Going forward we should improve the steps as the util is in placed we should leverage it better. You may address this in the separate PR.
